language/testing-infra/diem-transactional-test-harness/src/diem_test_harness.rs

```diff
@@ -0,0 +1,395 @@
+        let currency_code_tag = type_tag_for_currency_code(balance_currency_code);
+        let balance_resource_tag = BalanceResource::struct_tag_for_currency(currency_code_tag);
+        let balance_access_path =
+            AccessPath::resource_access_path(ResourceKey::new(*signer_addr, balance_resource_tag));
+
+        let balance_blob = self
+            .storage
+            .get(&balance_access_path)
+            .unwrap()
+            .ok_or_else(|| {
+                format_err!(
+                    "Failed to fetch balance resource under address {}.",
+                    signer_addr
+                )
+            })?;
+
+        Ok(bcs::from_bytes(&balance_blob).unwrap())
+    }
+
+    /// Derive the default transaction parameters from the account and balance resources fetched
+    /// from storage. In the future, we are planning to allow the user to override these using
+    /// command arguments.
+    fn fetch_default_transaction_parameters(
+        &self,
+        signer_addr: &AccountAddress,
+    ) -> Result<TransactionParameters> {
+        let account_resource = self.fetch_account_resource(signer_addr)?;
+
+        let sequence_number = account_resource.sequence_number();
+        let gas_currency_code = XUS_NAME.to_string();
+        let max_number_of_gas_units = GasConstants::default().maximum_number_of_gas_units;
+        let gas_unit_price = 0;
+        let max_gas_amount = if gas_unit_price == 0 {
+            max_number_of_gas_units.get()
+        } else {
+            let account_balance = self.fetch_balance_resource(
+                signer_addr,
+                account_config::from_currency_code_string(&gas_currency_code).unwrap(),
+            )?;
+            std::cmp::min(
+                max_number_of_gas_units.get(),
+                account_balance.coin() / gas_unit_price,
+            )
+        };
+
+        Ok(TransactionParameters {
+            sequence_number,
+            gas_currency_code,
+            gas_unit_price,
+            max_gas_amount,
+            expiration_timestamp_secs: 40000,
+        })
+    }
+
+    /// Perform a single Diem transaction.
+    ///
+    /// Should error if the transaction ends up being discarded, or having a status other than
+    /// EXECUTED.
+    fn run_transaction(&mut self, txn: SignedTransaction) -> Result<()> {
+        let mut outputs = DiemVM::execute_block_and_keep_vm_status(
+            vec![Transaction::UserTransaction(txn)],
+            &self.storage,
+        )
+        .unwrap();
+
+        assert_eq!(outputs.len(), 1);
+
+        let (status, output) = outputs.pop().unwrap();
+        match output.status() {
+            TransactionStatus::Keep(kept_vm_status) => match kept_vm_status {
+                KeptVMStatus::Executed => {
+                    self.storage.add_write_set(output.write_set());
+                }
+                _ => {
+                    bail!("Failed to execute transaction. VMStatus: {}", status)
```

Comment:
We might want to change the printing here slightly (I had to for the VMError stuff). but it is fine for now, and very easy to change